packages/nimble-components/src/table/models/keyboard-navigation-manager.ts

```diff
@@ -0,0 +1,1025 @@
+import {
+    TableFocusType,
+    TableRowFocusableElements,
+    type TableRecord
+} from '../types';
+import type { Virtualizer } from './virtualizer';
+import { TableGroupRow } from '../components/group-row';
+import { TableRow } from '../components/row';
+import { TableCell } from '../components/cell';
+import { MenuButton } from '../../menu-button';
+import { tableHeaderTag } from '../components/header';
+import { MenuItem } from '../../menu-item';
+import { Menu } from '../../menu';
+import type { TabIndexOverride } from '../../patterns/tab-index-override/types';
+import { TableCellView } from '../../table-column/base/cell-view';
+
+interface TableFocusState {
+    focusType: TableFocusType;
+    headerActionIndex?: number;
+    rowIndex?: number;
+    columnIndex?: number;
+    cellContentIndex?: number;
+}
+
+interface TableHeaderFocusableElements {
+    headerActions: HTMLElement[];
+    columnHeaders: HTMLElement[];
+}
+
+/**
+ * Manages the keyboard navigation and focus within the table.
+ * @internal
+ */
+export class KeyboardNavigationManager<TData extends TableRecord>
+implements Subscriber {
+    private readonly focusState: TableFocusState = {
+        focusType: TableFocusType.none
+    };
+
+    private readonly tableNotifier: Notifier;
+    private focusableHeaderElements!: TableHeaderFocusableElements;
+    private inNavigationMode = true;
+    private handlingTabPress = false;
+
+    public constructor(
+        private readonly table: Table<TData>,
+        private readonly virtualizer: Virtualizer<TData>
+    ) {
+        table.addEventListener('keydown', e => this.onCaptureKeyDown(e), {
+            capture: true
+        });
+        table.addEventListener('keydown', e => this.onKeyDown(e));
+        table.addEventListener('focusin', e => this.handleFocus(e));
+        table.addEventListener('focusout', e => {
+            console.log(
+                'table focusout',
+                'target',
+                e.target,
+                'relatedTarget',
+                e.relatedTarget,
+                'active element',
+                this.getActiveElementDebug()
+            );
+        });
+        table.addEventListener('blur', e => {
+            console.log(
+                'table blur',
+                'target',
+                e.target,
+                'relatedTarget',
+                e.relatedTarget,
+                'active element',
+                this.getActiveElementDebug()
+            );
+        });
+        this.tableNotifier = Observable.getNotifier(this.table);
+        this.tableNotifier.subscribe(this, 'rowElements');
+    }
+
+    public handleChange(source: unknown, args: unknown): void {
+        if (
+            source === this.table
+            && args === 'rowElements'
+            && this.hasRowOrCellFocusType()
+            && this.inNavigationMode
+        ) {
+            this.focusCurrentRow(false);
+        }
+    }
+
+    public onVirtualizerChange(): void {
```

Comment:
This is how we handle re-focusing the appropriate row after scrolling / virtualizer changes. Observing the `virtualizer.visibleItems` is another option, but we'd also need to observe each individual row and monitor for `dataIndex` changes.  
Possibly keeping a reference to the currently focused row (so we could just watch for `dataIndex` changes) is another option.

Reply:
Noticed while demoing to Milan / Meyer today that this simplified approach has issues once you start scrolling down with just DownArrow (you end up with no focused row visible).
Adding a rAF works but is hacky, so I'll probably switch back to the original approach mentioned in the comment above.